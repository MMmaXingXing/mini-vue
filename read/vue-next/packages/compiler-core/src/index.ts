export { baseCompile } from './compile'

// Also expose lower level APIs & types
export {
  CompilerOptions,
  ParserOptions,
  TransformOptions,
  CodegenOptions,
  HoistTransform,
  BindingMetadata,
  BindingTypes
} from './options'
export { baseParse, TextModes } from './parse'
export {
  transform,
  TransformContext,
  createTransformContext,
  traverseNode,
  createStructuralDirectiveTransform,
  NodeTransform,
  StructuralDirectiveTransform,
  DirectiveTransform
} from './transform'
export { generate, CodegenContext, CodegenResult } from './codegen'
export {
  ErrorCodes,
  CoreCompilerError,
  CompilerError,
  createCompilerError
} from './errors'

export * from './ast'
export * from './utils'
export * from './babelUtils'
export * from './runtimeHelpers'

export { getBaseTransformPreset, TransformPreset } from './compile'
export { transformModel } from './transforms/vModel'
export { transformOn } from './transforms/vOn'
export { transformBind } from './transforms/vBind'
export { noopDirectiveTransform } from './transforms/noopDirectiveTransform'
export { processIf } from './transforms/vIf'
export { processFor, createForLoopParams } from './transforms/vFor'
export {
  transformExpression,
  processExpression
} from './transforms/transformExpression'
export {
  buildSlots,
  SlotFnBuilder,
  trackVForSlotScopes,
  trackSlotScopes
} from './transforms/vSlot'
export {
  transformElement,
  resolveComponentType,
  buildProps
} from './transforms/transformElement'
export { processSlotOutlet } from './transforms/transformSlotOutlet'
export { generateCodeFrame } from '@vue/shared'

// v2 compat only
export {
  checkCompatEnabled,
  warnDeprecation,
  CompilerDeprecationTypes
} from './compat/compatConfig'
