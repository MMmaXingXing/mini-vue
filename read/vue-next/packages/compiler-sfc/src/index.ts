// API
export { parse } from './parse'
export { compileTemplate } from './compileTemplate'
export { compileStyle, compileStyleAsync } from './compileStyle'
export { compileScript } from './compileScript'
export { rewriteDefault } from './rewriteDefault'
export {
  shouldTransform as shouldTransformRef,
  transform as transformRef,
  transformAST as transformRefAST
} from '@vue/reactivity-transform'

// Utilities
export { parse as babelParse } from '@babel/parser'
import MagicString from 'magic-string'
export { MagicString }
// technically internal but we want it in @vue/repl, cast it as any to avoid
// relying on estree types
import { walk as _walk } from 'estree-walker'
export const walk = _walk as any
export {
  generateCodeFrame,
  walkIdentifiers,
  extractIdentifiers,
  isInDestructureAssignment,
  isStaticProperty
} from '@vue/compiler-core'

// Types
export {
  SFCParseOptions,
  SFCDescriptor,
  SFCBlock,
  SFCTemplateBlock,
  SFCScriptBlock,
  SFCStyleBlock,
  SFCParseResult
} from './parse'
export {
  TemplateCompiler,
  SFCTemplateCompileOptions,
  SFCTemplateCompileResults
} from './compileTemplate'
export {
  SFCStyleCompileOptions,
  SFCAsyncStyleCompileOptions,
  SFCStyleCompileResults
} from './compileStyle'
export { SFCScriptCompileOptions } from './compileScript'
export { AssetURLOptions, AssetURLTagConfig } from './templateTransformAssetUrl'
export {
  CompilerOptions,
  CompilerError,
  BindingMetadata
} from '@vue/compiler-core'
