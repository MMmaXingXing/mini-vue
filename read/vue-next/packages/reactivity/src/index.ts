export {
  ref,
  shallowRef,
  isRef,
  toRef,
  toRefs,
  unref,
  proxyRefs,
  customRef,
  triggerRef,
  Ref,
  ToRef,
  ToRefs,
  UnwrapRef,
  ShallowRef,
  ShallowUnwrapRef,
  RefUnwrapBailTypes,
  CustomRefFactory
} from './ref'
export {
  reactive,
  readonly,
  isReactive,
  isReadonly,
  isProxy,
  shallowReactive,
  shallowReadonly,
  markRaw,
  toRaw,
  ReactiveFlags,
  DeepReadonly,
  ShallowReactive,
  UnwrapNestedRefs
} from './reactive'
export {
  computed,
  ComputedRef,
  WritableComputedRef,
  WritableComputedOptions,
  ComputedGetter,
  ComputedSetter
} from './computed'
export { deferredComputed } from './deferredComputed'
export {
  effect,
  stop,
  trigger,
  track,
  enableTracking,
  pauseTracking,
  resetTracking,
  ITERATE_KEY,
  ReactiveEffect,
  ReactiveEffectRunner,
  ReactiveEffectOptions,
  EffectScheduler,
  DebuggerOptions,
  DebuggerEvent,
  DebuggerEventExtraInfo
} from './effect'
export {
  effectScope,
  EffectScope,
  getCurrentScope,
  onScopeDispose
} from './effectScope'
export { TrackOpTypes, TriggerOpTypes } from './operations'
