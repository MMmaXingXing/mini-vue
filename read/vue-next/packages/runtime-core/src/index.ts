// Core API ------------------------------------------------------------------

export const version = __VERSION__
export {
  // core
  computed,
  reactive,
  ref,
  readonly,
  // utilities
  unref,
  proxyRefs,
  isRef,
  toRef,
  toRefs,
  isProxy,
  isReactive,
  isReadonly,
  // advanced
  customRef,
  triggerRef,
  shallowRef,
  shallowReactive,
  shallowReadonly,
  markRaw,
  toRaw,
  // effect
  effect,
  stop,
  ReactiveEffect,
  // effect scope
  effectScope,
  EffectScope,
  getCurrentScope,
  onScopeDispose
} from '@vue/reactivity'
export {
  watch,
  watchEffect,
  watchPostEffect,
  watchSyncEffect
} from './apiWatch'
export {
  onBeforeMount,
  onMounted,
  onBeforeUpdate,
  onUpdated,
  onBeforeUnmount,
  onUnmounted,
  onActivated,
  onDeactivated,
  onRenderTracked,
  onRenderTriggered,
  onErrorCaptured,
  onServerPrefetch
} from './apiLifecycle'
export { provide, inject } from './apiInject'
export { nextTick } from './scheduler'
export { defineComponent } from './apiDefineComponent'
export { defineAsyncComponent } from './apiAsyncComponent'
export { useAttrs, useSlots } from './apiSetupHelpers'

// <script setup> API ----------------------------------------------------------

export {
  // macros runtime, for typing and warnings only
  defineProps,
  defineEmits,
  defineExpose,
  withDefaults,
  // internal
  mergeDefaults,
  createPropsRestProxy,
  withAsyncContext
} from './apiSetupHelpers'

// Advanced API ----------------------------------------------------------------

// For getting a hold of the internal instance in setup() - useful for advanced
// plugins
export { getCurrentInstance } from './component'

// For raw render function users
export { h } from './h'
// Advanced render function utilities
export { createVNode, cloneVNode, mergeProps, isVNode } from './vnode'
// VNode types
export { Fragment, Text, Comment, Static } from './vnode'
// Built-in components
export { Teleport, TeleportProps } from './components/Teleport'
export { Suspense, SuspenseProps } from './components/Suspense'
export { KeepAlive, KeepAliveProps } from './components/KeepAlive'
export {
  BaseTransition,
  BaseTransitionProps
} from './components/BaseTransition'
// For using custom directives
export { withDirectives } from './directives'
// SSR context
export { useSSRContext, ssrContextKey } from './helpers/useSsrContext'

// Custom Renderer API ---------------------------------------------------------

export { createRenderer, createHydrationRenderer } from './renderer'
export { queuePostFlushCb } from './scheduler'
export { warn } from './warning'
export {
  handleError,
  callWithErrorHandling,
  callWithAsyncErrorHandling,
  ErrorCodes
} from './errorHandling'
export {
  resolveComponent,
  resolveDirective,
  resolveDynamicComponent
} from './helpers/resolveAssets'
// For integration with runtime compiler
export { registerRuntimeCompiler, isRuntimeOnly } from './component'
export {
  useTransitionState,
  resolveTransitionHooks,
  setTransitionHooks,
  getTransitionRawChildren
} from './components/BaseTransition'
export { initCustomFormatter } from './customFormatter'

// For devtools
export { devtools, setDevtoolsHook } from './devtools'

// Types -------------------------------------------------------------------------

import { VNode } from './vnode'
import { ComponentInternalInstance } from './component'

// Augment Ref unwrap bail types.
// Note: if updating this, also update `types/refBail.d.ts`.
declare module '@vue/reactivity' {
  export interface RefUnwrapBailTypes {
    runtimeCoreBailTypes:
      | VNode
      | {
          // directly bailing on ComponentPublicInstance results in recursion
          // so we use this as a bail hint
          $: ComponentInternalInstance
        }
  }
}

export {
  Ref,
  ToRef,
  ToRefs,
  UnwrapRef,
  ShallowRef,
  ShallowUnwrapRef,
  CustomRefFactory,
  ReactiveFlags,
  DeepReadonly,
  ShallowReactive,
  UnwrapNestedRefs,
  ComputedRef,
  WritableComputedRef,
  WritableComputedOptions,
  ComputedGetter,
  ComputedSetter,
  ReactiveEffectRunner,
  ReactiveEffectOptions,
  EffectScheduler,
  DebuggerOptions,
  DebuggerEvent,
  DebuggerEventExtraInfo,
  TrackOpTypes,
  TriggerOpTypes
} from '@vue/reactivity'
export {
  WatchEffect,
  WatchOptions,
  WatchOptionsBase,
  WatchCallback,
  WatchSource,
  WatchStopHandle
} from './apiWatch'
export { InjectionKey } from './apiInject'
export {
  App,
  AppConfig,
  AppContext,
  Plugin,
  CreateAppFunction,
  OptionMergeFunction
} from './apiCreateApp'
export {
  VNode,
  VNodeChild,
  VNodeTypes,
  VNodeProps,
  VNodeArrayChildren,
  VNodeNormalizedChildren
} from './vnode'
export {
  Component,
  ConcreteComponent,
  FunctionalComponent,
  ComponentInternalInstance,
  SetupContext,
  ComponentCustomProps,
  AllowedComponentProps
} from './component'
export { DefineComponent } from './apiDefineComponent'
export {
  ComponentOptions,
  ComponentOptionsMixin,
  ComponentOptionsWithoutProps,
  ComponentOptionsWithObjectProps,
  ComponentOptionsWithArrayProps,
  ComponentCustomOptions,
  ComponentOptionsBase,
  RenderFunction,
  MethodOptions,
  ComputedOptions,
  RuntimeCompilerOptions
} from './componentOptions'
export { EmitsOptions, ObjectEmitsOptions } from './componentEmits'
export {
  ComponentPublicInstance,
  ComponentCustomProperties,
  CreateComponentPublicInstance
} from './componentPublicInstance'
export {
  Renderer,
  RendererNode,
  RendererElement,
  HydrationRenderer,
  RendererOptions,
  RootRenderFunction
} from './renderer'
export { RootHydrateFunction } from './hydration'
export { Slot, Slots } from './componentSlots'
export {
  Prop,
  PropType,
  ComponentPropsOptions,
  ComponentObjectPropsOptions,
  ExtractPropTypes,
  ExtractDefaultPropTypes
} from './componentProps'
export {
  Directive,
  DirectiveBinding,
  DirectiveHook,
  ObjectDirective,
  FunctionDirective,
  DirectiveArguments
} from './directives'
export { SuspenseBoundary } from './components/Suspense'
export { TransitionState, TransitionHooks } from './components/BaseTransition'
export {
  AsyncComponentOptions,
  AsyncComponentLoader
} from './apiAsyncComponent'
export { HMRRuntime } from './hmr'

// Internal API ----------------------------------------------------------------

// **IMPORTANT** Internal APIs may change without notice between versions and
// user code should avoid relying on them.

// For compiler generated code
// should sync with '@vue/compiler-core/src/runtimeHelpers.ts'
export {
  withCtx,
  pushScopeId,
  popScopeId,
  withScopeId
} from './componentRenderContext'
export { renderList } from './helpers/renderList'
export { toHandlers } from './helpers/toHandlers'
export { renderSlot } from './helpers/renderSlot'
export { createSlots } from './helpers/createSlots'
export { withMemo, isMemoSame } from './helpers/withMemo'
export {
  openBlock,
  createBlock,
  setBlockTracking,
  createTextVNode,
  createCommentVNode,
  createStaticVNode,
  createElementVNode,
  createElementBlock,
  guardReactiveProps
} from './vnode'
export {
  toDisplayString,
  camelize,
  capitalize,
  toHandlerKey,
  normalizeProps,
  normalizeClass,
  normalizeStyle
} from '@vue/shared'

// For test-utils
export { transformVNodeArgs } from './vnode'

// SSR -------------------------------------------------------------------------

// **IMPORTANT** These APIs are exposed solely for @vue/server-renderer and may
// change without notice between versions. User code should never rely on them.

import { createComponentInstance, setupComponent } from './component'
import { renderComponentRoot } from './componentRenderUtils'
import { setCurrentRenderingInstance } from './componentRenderContext'
import { isVNode, normalizeVNode } from './vnode'

const _ssrUtils = {
  createComponentInstance,
  setupComponent,
  renderComponentRoot,
  setCurrentRenderingInstance,
  isVNode,
  normalizeVNode
}

/**
 * SSR utils for \@vue/server-renderer. Only exposed in cjs builds.
 * @internal
 */
export const ssrUtils = (__SSR__ ? _ssrUtils : null) as typeof _ssrUtils

// 2.x COMPAT ------------------------------------------------------------------

export { DeprecationTypes } from './compat/compatConfig'
export { CompatVue } from './compat/global'
export { LegacyConfig } from './compat/globalConfig'

import { warnDeprecation } from './compat/compatConfig'
import { createCompatVue } from './compat/global'
import {
  isCompatEnabled,
  checkCompatEnabled,
  softAssertCompatEnabled
} from './compat/compatConfig'
import { resolveFilter as _resolveFilter } from './helpers/resolveAssets'

/**
 * @internal only exposed in compat builds
 */
export const resolveFilter = __COMPAT__ ? _resolveFilter : null

const _compatUtils = {
  warnDeprecation,
  createCompatVue,
  isCompatEnabled,
  checkCompatEnabled,
  softAssertCompatEnabled
}

/**
 * @internal only exposed in compat builds.
 */
export const compatUtils = (
  __COMPAT__ ? _compatUtils : null
) as typeof _compatUtils
