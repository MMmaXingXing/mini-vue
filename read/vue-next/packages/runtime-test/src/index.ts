import {
  createRenderer,
  VNode,
  RootRenderFunction,
  CreateAppFunction
} from '@vue/runtime-core'
import { nodeOps, TestElement } from './nodeOps'
import { patchProp } from './patchProp'
import { serializeInner } from './serialize'
import { extend } from '@vue/shared'

const { render: baseRender, createApp: baseCreateApp } = createRenderer(
  extend({ patchProp }, nodeOps)
)

export const render = baseRender as RootRenderFunction<TestElement>
export const createApp = baseCreateApp as CreateAppFunction<TestElement>

// convenience for one-off render validations
export function renderToString(vnode: VNode) {
  const root = nodeOps.createElement('div')
  render(vnode, root)
  return serializeInner(root)
}

export * from './triggerEvent'
export * from './serialize'
export * from './nodeOps'
export * from '@vue/runtime-core'
