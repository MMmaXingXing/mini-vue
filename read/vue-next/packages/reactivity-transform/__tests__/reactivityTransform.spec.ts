import { parse } from '@babel/parser'
import { transform } from '../src'

function assertCode(code: string) {
  // parse the generated code to make sure it is valid
  try {
    parse(code, {
      sourceType: 'module',
      plugins: ['typescript']
    })
  } catch (e: any) {
    console.log(code)
    throw e
  }
  expect(code).toMatchSnapshot()
}

test('$ unwrapping', () => {
  const { code, rootRefs } = transform(`
    import { ref, shallowRef } from 'vue'
    let foo = $(ref())
    let a = $(ref(1))
    let b = $(shallowRef({
      count: 0
    }))
    let c = () => {}
    let d
    `)
  expect(code).not.toMatch(`$(ref())`)
  expect(code).not.toMatch(`$(ref(1))`)
  expect(code).not.toMatch(`$(shallowRef({`)
  expect(code).toMatch(`let foo = (ref())`)
  expect(code).toMatch(`let a = (ref(1))`)
  expect(code).toMatch(`
    let b = (shallowRef({
      count: 0
    }))
    `)
  // normal declarations left untouched
  expect(code).toMatch(`let c = () => {}`)
  expect(code).toMatch(`let d`)
  expect(rootRefs).toStrictEqual(['foo', 'a', 'b'])
  assertCode(code)
})

test('$ref & $shallowRef declarations', () => {
  const { code, rootRefs, importedHelpers } = transform(`
    let foo = $ref()
    let a = $ref(1)
    let b = $shallowRef({
      count: 0
    })
    let c = () => {}
    let d
    `)
  expect(code).toMatch(
    `import { ref as _ref, shallowRef as _shallowRef } from 'vue'`
  )
  expect(code).not.toMatch(`$ref()`)
  expect(code).not.toMatch(`$ref(1)`)
  expect(code).not.toMatch(`$shallowRef({`)
  expect(code).toMatch(`let foo = _ref()`)
  expect(code).toMatch(`let a = _ref(1)`)
  expect(code).toMatch(`
    let b = _shallowRef({
      count: 0
    })
    `)
  // normal declarations left untouched
  expect(code).toMatch(`let c = () => {}`)
  expect(code).toMatch(`let d`)
  expect(rootRefs).toStrictEqual(['foo', 'a', 'b'])
  expect(importedHelpers).toStrictEqual(['ref', 'shallowRef'])
  assertCode(code)
})

test('multi $ref declarations', () => {
  const { code, rootRefs, importedHelpers } = transform(`
    let a = $ref(1), b = $ref(2), c = $ref({
      count: 0
    })
    `)
  expect(code).toMatch(`
    let a = _ref(1), b = _ref(2), c = _ref({
      count: 0
    })
    `)
  expect(rootRefs).toStrictEqual(['a', 'b', 'c'])
  expect(importedHelpers).toStrictEqual(['ref'])
  assertCode(code)
})

test('$computed declaration', () => {
  const { code, rootRefs, importedHelpers } = transform(`
    let a = $computed(() => 1)
    `)
  expect(code).toMatch(`
    let a = _computed(() => 1)
    `)
  expect(rootRefs).toStrictEqual(['a'])
  expect(importedHelpers).toStrictEqual(['computed'])
  assertCode(code)
})

test('mixing $ref & $computed declarations', () => {
  const { code, rootRefs, importedHelpers } = transform(`
    let a = $ref(1), b = $computed(() => a + 1)
    `)
  expect(code).toMatch(`
    let a = _ref(1), b = _computed(() => a.value + 1)
    `)
  expect(rootRefs).toStrictEqual(['a', 'b'])
  expect(importedHelpers).toStrictEqual(['ref', 'computed'])
  assertCode(code)
})

test('accessing ref binding', () => {
  const { code } = transform(`
    let a = $ref(1)
    console.log(a)
    function get() {
      return a + 1
    }
    `)
  expect(code).toMatch(`console.log(a.value)`)
  expect(code).toMatch(`return a.value + 1`)
  assertCode(code)
})

describe('cases that should not append .value', () => {
  test('member expression', () => {
    const { code } = transform(`
      let a = $ref(1)
      console.log(b.a)
      `)
    expect(code).not.toMatch(`a.value`)
  })

  test('function argument', () => {
    const { code } = transform(`
      let a = $ref(1)
      function get(a) {
        return a + 1
      }
      function get2({ a }) {
        return a + 1
      }
      function get3([a]) {
        return a + 1
      }
      `)
    expect(code).not.toMatch(`a.value`)
  })

  test('for in/of loops', () => {
    const { code } = transform(`
    let a = $ref(1)
    for (const [a, b] of arr) {
      console.log(a)
    }
    for (let a in arr) {
      console.log(a)
    }
    `)
    expect(code).not.toMatch(`a.value`)
  })
})

test('mutating ref binding', () => {
  const { code } = transform(`
    let a = $ref(1)
    let b = $ref({ count: 0 })
    function inc() {
      a++
      a = a + 1
      b.count++
      b.count = b.count + 1
      ;({ a } = { a: 2 })
      ;[a] = [1]
    }
    `)
  expect(code).toMatch(`a.value++`)
  expect(code).toMatch(`a.value = a.value + 1`)
  expect(code).toMatch(`b.value.count++`)
  expect(code).toMatch(`b.value.count = b.value.count + 1`)
  expect(code).toMatch(`;({ a: a.value } = { a: 2 })`)
  expect(code).toMatch(`;[a.value] = [1]`)
  assertCode(code)
})

test('using ref binding in property shorthand', () => {
  const { code } = transform(`
    let a = $ref(1)
    const b = { a }
    function test() {
      const { a } = b
    }
    `)
  expect(code).toMatch(`const b = { a: a.value }`)
  // should not convert destructure
  expect(code).toMatch(`const { a } = b`)
  assertCode(code)
})

test('should not rewrite scope variable', () => {
  const { code } = transform(`

      let a = $ref(1)
      let b = $ref(1)
      let d = $ref(1)
      const e = 1
      function test() {
        const a = 2
        console.log(a)
        console.log(b)
        let c = { c: 3 }
        console.log(c)
        console.log(d)
        console.log(e)
      }
    `)
  expect(code).toMatch('console.log(a)')
  expect(code).toMatch('console.log(b.value)')
  expect(code).toMatch('console.log(c)')
  expect(code).toMatch('console.log(d.value)')
  expect(code).toMatch('console.log(e)')
  assertCode(code)
})

test('object destructure', () => {
  const { code, rootRefs } = transform(`
    let n = $ref(1), { a, b: c, d = 1, e: f = 2, [g]: h } = $(useFoo())
    let { foo } = $(useSomthing(() => 1));
    console.log(n, a, c, d, f, h, foo)
    `)
  expect(code).toMatch(`a = _toRef(__$temp_1, 'a')`)
  expect(code).toMatch(`c = _toRef(__$temp_1, 'b')`)
  expect(code).toMatch(`d = _toRef(__$temp_1, 'd', 1)`)
  expect(code).toMatch(`f = _toRef(__$temp_1, 'e', 2)`)
  expect(code).toMatch(`h = _toRef(__$temp_1, g)`)
  expect(code).toMatch(`foo = _toRef(__$temp_2, 'foo')`)
  expect(code).toMatch(
    `console.log(n.value, a.value, c.value, d.value, f.value, h.value, foo.value)`
  )
  expect(rootRefs).toStrictEqual(['n', 'a', 'c', 'd', 'f', 'h', 'foo'])
  assertCode(code)
})

test('object destructure w/ mid-path default values', () => {
  const { code, rootRefs } = transform(`
    const { a: { b } = { b: 123 }} = $(useFoo())
    console.log(b)
  `)
  expect(code).toMatch(`b = _toRef((__$temp_1.a || { b: 123 }), 'b')`)
  expect(code).toMatch(`console.log(b.value)`)
  expect(rootRefs).toStrictEqual(['b'])
  assertCode(code)
})

test('array destructure', () => {
  const { code, rootRefs } = transform(`
    let n = $ref(1), [a, b = 1] = $(useFoo())
    console.log(n, a, b)
    `)
  expect(code).toMatch(`a = _toRef(__$temp_1, 0)`)
  expect(code).toMatch(`b = _toRef(__$temp_1, 1, 1)`)
  expect(code).toMatch(`console.log(n.value, a.value, b.value)`)
  expect(rootRefs).toStrictEqual(['n', 'a', 'b'])
  assertCode(code)
})

test('nested destructure', () => {
  const { code, rootRefs } = transform(`
    let [{ a: { b }}] = $(useFoo())
    let { c: [d, e] } = $(useBar())
    console.log(b, d, e)
    `)
  expect(code).toMatch(`b = _toRef(__$temp_1[0].a, 'b')`)
  expect(code).toMatch(`d = _toRef(__$temp_2.c, 0)`)
  expect(code).toMatch(`e = _toRef(__$temp_2.c, 1)`)
  expect(rootRefs).toStrictEqual(['b', 'd', 'e'])
  assertCode(code)
})

test('$$', () => {
  const { code } = transform(`
    let a = $ref(1)
    const b = $$(a)
    const c = $$({ a })
    callExternal($$(a))
    `)
  expect(code).toMatch(`const b = (a)`)
  expect(code).toMatch(`const c = ({ a })`)
  expect(code).toMatch(`callExternal((a))`)
  assertCode(code)
})

test('nested scopes', () => {
  const { code, rootRefs } = transform(`
    let a = $ref(0)
    let b = $ref(0)
    let c = 0

    a++ // outer a
    b++ // outer b
    c++ // outer c

    let bar = $ref(0)
    bar++ // outer bar

    function foo({ a }) {
      a++ // inner a
      b++ // inner b
      let c = $ref(0)
      c++ // inner c
      let d = $ref(0)

      function bar(c) {
        c++ // nested c
        d++ // nested d
      }
      bar() // inner bar

      if (true) {
        let a = $ref(0)
        a++ // if block a
      }

      return $$({ a, b, c, d })
    }
    `)
  expect(rootRefs).toStrictEqual(['a', 'b', 'bar'])

  expect(code).toMatch('a.value++ // outer a')
  expect(code).toMatch('b.value++ // outer b')
  expect(code).toMatch('c++ // outer c')

  expect(code).toMatch('a++ // inner a') // shadowed by function arg
  expect(code).toMatch('b.value++ // inner b')
  expect(code).toMatch('c.value++ // inner c') // shadowed by local ref binding

  expect(code).toMatch('c++ // nested c') // shadowed by inline fn arg
  expect(code).toMatch(`d.value++ // nested d`)

  expect(code).toMatch(`a.value++ // if block a`) // if block

  expect(code).toMatch(`bar.value++ // outer bar`)
  // inner bar shadowed by function declaration
  expect(code).toMatch(`bar() // inner bar`)

  expect(code).toMatch(`return ({ a, b, c, d })`)
  assertCode(code)
})

//#4062
test('should not rewrite type identifiers', () => {
  const { code } = transform(
    `const props = defineProps<{msg: string; ids?: string[]}>()
        let ids = $ref([])`,
    {
      parserPlugins: ['typescript']
    }
  )
  expect(code).not.toMatch('.value')
  assertCode(code)
})

// #4254
test('handle TS casting syntax', () => {
  const { code } = transform(
    `
      let a = $ref(1)
      console.log(a!)
      console.log(a! + 1)
      console.log(a as number)
      console.log((a as number) + 1)
      console.log(<number>a)
      console.log(<number>a + 1)
      console.log(a! + (a as number))
      console.log(a! + <number>a)
      console.log((a as number) + <number>a)
      `,
    {
      parserPlugins: ['typescript']
    }
  )
  expect(code).toMatch('console.log(a.value!)')
  expect(code).toMatch('console.log(a.value as number)')
  expect(code).toMatch('console.log(<number>a.value)')
  assertCode(code)
})

test('macro import alias and removal', () => {
  const { code } = transform(
    `
    import { $ as fromRefs, $ref } from 'vue/macros'

    let a = $ref(1)
    const { x, y } = fromRefs(useMouse())
    `
  )
  // should remove imports
  expect(code).not.toMatch(`from 'vue/macros'`)
  expect(code).toMatch(`let a = _ref(1)`)
  expect(code).toMatch(`const __$temp_1 = (useMouse())`)
  assertCode(code)
})

describe('errors', () => {
  test('$ref w/ destructure', () => {
    expect(() => transform(`let { a } = $ref(1)`)).toThrow(
      `cannot be used with destructure`
    )
  })

  test('$computed w/ destructure', () => {
    expect(() => transform(`let { a } = $computed(() => 1)`)).toThrow(
      `cannot be used with destructure`
    )
  })

  test('warn usage in non-init positions', () => {
    expect(() =>
      transform(
        `let bar = $ref(1)
          bar = $ref(2)`
      )
    ).toThrow(`$ref can only be used as the initializer`)

    expect(() => transform(`let bar = { foo: $computed(1) }`)).toThrow(
      `$computed can only be used as the initializer`
    )
  })

  test('not transform the prototype attributes', () => {
    const { code } = transform(`
    const hasOwnProperty = Object.prototype.hasOwnProperty
    const hasOwn = (val, key) => hasOwnProperty.call(val, key)
    `)
    expect(code).not.toMatch('.value')
  })

  test('rest element in $() destructure', () => {
    expect(() => transform(`let { a, ...b } = $(foo())`)).toThrow(
      `does not support rest element`
    )
    expect(() => transform(`let [a, ...b] = $(foo())`)).toThrow(
      `does not support rest element`
    )
  })
})
