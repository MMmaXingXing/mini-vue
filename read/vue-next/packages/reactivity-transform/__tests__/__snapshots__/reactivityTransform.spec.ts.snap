// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$ unwrapping 1`] = `
"
    import { ref, shallowRef } from 'vue'
    let foo = (ref())
    let a = (ref(1))
    let b = (shallowRef({
      count: 0
    }))
    let c = () => {}
    let d
    "
`;

exports[`$$ 1`] = `
"import { ref as _ref } from 'vue'

    let a = _ref(1)
    const b = (a)
    const c = ({ a })
    callExternal((a))
    "
`;

exports[`$computed declaration 1`] = `
"import { computed as _computed } from 'vue'

    let a = _computed(() => 1)
    "
`;

exports[`$ref & $shallowRef declarations 1`] = `
"import { ref as _ref, shallowRef as _shallowRef } from 'vue'

    let foo = _ref()
    let a = _ref(1)
    let b = _shallowRef({
      count: 0
    })
    let c = () => {}
    let d
    "
`;

exports[`accessing ref binding 1`] = `
"import { ref as _ref } from 'vue'

    let a = _ref(1)
    console.log(a.value)
    function get() {
      return a.value + 1
    }
    "
`;

exports[`array destructure 1`] = `
"import { ref as _ref, toRef as _toRef } from 'vue'

    let n = _ref(1), __$temp_1 = (useFoo()),
  a = _toRef(__$temp_1, 0),
  b = _toRef(__$temp_1, 1, 1)
    console.log(n.value, a.value, b.value)
    "
`;

exports[`handle TS casting syntax 1`] = `
"import { ref as _ref } from 'vue'

      let a = _ref(1)
      console.log(a.value!)
      console.log(a.value! + 1)
      console.log(a.value as number)
      console.log((a.value as number) + 1)
      console.log(<number>a.value)
      console.log(<number>a.value + 1)
      console.log(a.value! + (a.value as number))
      console.log(a.value! + <number>a.value)
      console.log((a.value as number) + <number>a.value)
      "
`;

exports[`macro import alias and removal 1`] = `
"import { ref as _ref, toRef as _toRef } from 'vue'

    

    let a = _ref(1)
    const __$temp_1 = (useMouse()),
  x = _toRef(__$temp_1, 'x'),
  y = _toRef(__$temp_1, 'y')
    "
`;

exports[`mixing $ref & $computed declarations 1`] = `
"import { ref as _ref, computed as _computed } from 'vue'

    let a = _ref(1), b = _computed(() => a.value + 1)
    "
`;

exports[`multi $ref declarations 1`] = `
"import { ref as _ref } from 'vue'

    let a = _ref(1), b = _ref(2), c = _ref({
      count: 0
    })
    "
`;

exports[`mutating ref binding 1`] = `
"import { ref as _ref } from 'vue'

    let a = _ref(1)
    let b = _ref({ count: 0 })
    function inc() {
      a.value++
      a.value = a.value + 1
      b.value.count++
      b.value.count = b.value.count + 1
      ;({ a: a.value } = { a: 2 })
      ;[a.value] = [1]
    }
    "
`;

exports[`nested destructure 1`] = `
"import { toRef as _toRef } from 'vue'

    let __$temp_1 = (useFoo()),
  b = _toRef(__$temp_1[0].a, 'b')
    let __$temp_2 = (useBar()),
  d = _toRef(__$temp_2.c, 0),
  e = _toRef(__$temp_2.c, 1)
    console.log(b.value, d.value, e.value)
    "
`;

exports[`nested scopes 1`] = `
"import { ref as _ref } from 'vue'

    let a = _ref(0)
    let b = _ref(0)
    let c = 0

    a.value++ // outer a
    b.value++ // outer b
    c++ // outer c

    let bar = _ref(0)
    bar.value++ // outer bar

    function foo({ a }) {
      a++ // inner a
      b.value++ // inner b
      let c = _ref(0)
      c.value++ // inner c
      let d = _ref(0)

      function bar(c) {
        c++ // nested c
        d.value++ // nested d
      }
      bar() // inner bar

      if (true) {
        let a = _ref(0)
        a.value++ // if block a
      }

      return ({ a, b, c, d })
    }
    "
`;

exports[`object destructure 1`] = `
"import { ref as _ref, toRef as _toRef } from 'vue'

    let n = _ref(1), __$temp_1 = (useFoo()),
  a = _toRef(__$temp_1, 'a'),
  c = _toRef(__$temp_1, 'b'),
  d = _toRef(__$temp_1, 'd', 1),
  f = _toRef(__$temp_1, 'e', 2),
  h = _toRef(__$temp_1, g)
    let __$temp_2 = (useSomthing(() => 1)),
  foo = _toRef(__$temp_2, 'foo');
    console.log(n.value, a.value, c.value, d.value, f.value, h.value, foo.value)
    "
`;

exports[`object destructure w/ mid-path default values 1`] = `
"import { toRef as _toRef } from 'vue'

    const __$temp_1 = (useFoo()),
  b = _toRef((__$temp_1.a || { b: 123 }), 'b')
    console.log(b.value)
  "
`;

exports[`should not rewrite scope variable 1`] = `
"import { ref as _ref } from 'vue'


      let a = _ref(1)
      let b = _ref(1)
      let d = _ref(1)
      const e = 1
      function test() {
        const a = 2
        console.log(a)
        console.log(b.value)
        let c = { c: 3 }
        console.log(c)
        console.log(d.value)
        console.log(e)
      }
    "
`;

exports[`should not rewrite type identifiers 1`] = `
"import { ref as _ref } from 'vue'
const props = defineProps<{msg: string; ids?: string[]}>()
        let ids = _ref([])"
`;

exports[`using ref binding in property shorthand 1`] = `
"import { ref as _ref } from 'vue'

    let a = _ref(1)
    const b = { a: a.value }
    function test() {
      const { a } = b
    }
    "
`;
